### ISSUE TITLE
optimize_for_inference.py should remove Dropout operations

### ISSUE THREAD

**Issue by mattrajca** [tensorflow/tensorflow#5867]
When I first tried using an exported MNIST model with TensorFlow on iOS, I got the following error:

    Invalid argument: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels:
      <no registered kernels>
    
         [[Node: dropout/random_uniform/RandomUniform = RandomUniform[T=DT_INT32, dtype=DT_FLOAT, seed=0, seed2=0](dropout/Shape)]]

Since Dropout operations are no-ops during inference (we pass in a keep probability of 1), it would be nice if they were removed (or turned into no-ops of some kind that can be parsed and ignored by the iOS library).

While I was able to work around this by explicitly exporting a separate graph that does not contain Dropout, it was pretty tedious and it would be nice if the `optimize_for_inference.py` script did this automatically.

### Environment info
Operating System: macOS 10.12

Installed version of CUDA and cuDNN: 
None

Source:
This week's tip-of-tree (around d93d526cb804896004c1c20a41586ce0e2415b9c)

**Comment by nikashitsa** [comment]
Yes, it would be nice if the `optimize_for_inference.py` script did this automatically.  
My current manual solution https://dato.ml/drop-dropout-from-frozen-model/

**Comment by cancan101** [comment]
related / dupe: https://github.com/tensorflow/tensorflow/issues/6124

**Comment by cancan101** [comment]
There is also this SO (attempt at a) solution: http://stackoverflow.com/questions/40358892/wipe-out-dropout-operations-from-tensorflow-graph

**Comment by swarathesh** [comment]
i'am getting the following error in android 
`ERROR: Analysis of target '//tensorflow/contrib/util:convert_graphdef_memmapped_format' failed; build aborted.
`

**Comment by haydenth** [comment]
Is there anyway this can be accomplished using the transform_graph tool and remove_nodes?

**Comment by kmonachopoulos** [comment]
I am using the optimize_for_inference script to eliminate the dropout nodes from my protobuf model. Before the script a part of the model has : 
```
Operation Name : dropout_1/keep_prob
Tensor Stats : (<tf.Tensor 'dropout_1/keep_prob:0' shape=() dtype=float32>,)
Operation Name : dropout_1/Shape
Tensor Stats : (<tf.Tensor 'dropout_1/Shape:0' shape=(4,) dtype=int32>,)
Operation Name : dropout_1/random_uniform/min
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform/min:0' shape=() dtype=float32>,)
Operation Name : dropout_1/random_uniform/max
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform/max:0' shape=() dtype=float32>,)
Operation Name : dropout_1/random_uniform/RandomUniform
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform/RandomUniform:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/random_uniform/sub
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform/sub:0' shape=() dtype=float32>,)
Operation Name : dropout_1/random_uniform/mul
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform/mul:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/random_uniform
Tensor Stats : (<tf.Tensor 'dropout_1/random_uniform:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/add
Tensor Stats : (<tf.Tensor 'dropout_1/add:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/Floor
Tensor Stats : (<tf.Tensor 'dropout_1/Floor:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/Inv
Tensor Stats : (<tf.Tensor 'dropout_1/Inv:0' shape=() dtype=float32>,)
Operation Name : dropout_1/mul
Tensor Stats : (<tf.Tensor 'dropout_1/mul:0' shape=(?, ?, ?, 4096) dtype=float32>,)
Operation Name : dropout_1/mul_1
Tensor Stats : (<tf.Tensor 'dropout_1/mul_1:0' shape=(?, ?, ?, 4096) dtype=float32>,)
```
I am using the script like this : 

```
bazel-bin/tensorflow/python/tools/optimize_for_inference \
--input=/home/konmon01/tensorflow/tensorflow/examples/KittiSeg/RUNS/KittiSeg_pretrained/Protobufs/FCN8_bilinear_session_Script_Freezed.pb \
--output=/home/konmon01/tensorflow/tensorflow/examples/KittiSeg/RUNS/KittiSeg_pretrained/Protobufs/FCN8_bilinear_session_Script_Freezed_Optimized.pb \
--frozen_graph=True \
--input_names=Inputs/fifo_queue_Dequeue \
--output_names=upscore32/conv2d_transpose
```
However after the implementation of the script the dropout layer is still there, exactly the same as above. I used other approaches as well but with no luck. Is there anything wrong in my implementation of this script?

**Comment by alextp** [comment]
Is using a tensor for keep_prob which is set to 1 during inference not enough?

**Comment by ngeorgis** [comment]
Any updates on this? Just tried optimize_for_inference.py but dropout still there confusing NVIDIA TensorRT and Jetson TX2 inference.

**Comment by kmonachopoulos** [comment]
There is a way to create a new graph and copy all the nodes you want (except dropout) from the old graph to the new one, fixing the input/output names of the nodes before and after the dropout layer. This way you can create a new graph without the dropout layer.

**Comment by AndRossi** [comment]
Please let optimize_for_inference remove dropout layers from the graph.
I am trying to convert a MNIST model from TF to TFLite.
It seems to me that this should be the very first thing to try when one is learning to use TFLite, since MNIST is the first tutorial in TensorFlow.

To allow the conversion to TFLite, I need to remove the dropout layer in my MNIST model frozen graph. 

If I get it right, the only way to do that (without writing a python script for that) is to use the Graph Transform Tool.
Which can only run with Bazel. 
Which requires to download TF from sources.
This feels like a **huge overkill**.

The funny thing is that in prediction mode the dropout would not even be used.
So I'm spending a day downloading and configuring components in my environment to just remove something that would not even be used.

If I am missing something please tell me - I would be overjoyed to know I got it wrong.

**Comment by shrutim90** [comment]
Run training script -> export_inference_graph -> freeze_graph . This removes all the extra nodes added during training. 

I was trying to convert finetuned inception model to uff format required by Jetson, and got issues due to the same RandomUniform node and a few others. Running export_inference_graph on finetuned model before freezing fixed the issue.

**Comment by shrutim90** [comment]
@AndRossi : In this link: https://github.com/tensorflow/models/tree/master/research/slim#Pretrained , they have used export_inference_graph script to export default inception graph. If you'll check the code for that script, you'll see there is an option to specify dataset_name as well. 

python export_inference_graph.py \
  --alsologtostderr \
  --model_name=inception_v3 \
  --dataset_name=flowers \
  --output_file=/tmp/inception_v3_inf_graph_flowers.pb

So you can specify any dataset predefined in slim library or add you own data in the slim dataset and then use above script, with required dataset_name. It will create graph with all the extra layers removed and correct number of output nodes.

**Comment by tofulawrence** [comment]
You should freeze the training checkpoint into a separate eval/inference graph. That graph should be compatible with TF Lite. 

In general, it's difficult to convert a training graph into an inference graph because batch norm and dropout create different graphs for training or eval.

**Comment by tofulawrence** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by kbalka** [comment]
@tofulawrence I bet you had a look on this https://github.com/tensorflow/tensorflow/pull/19264,  can you comment why it can't be upstreamed ? Does it fail some internal test cases ?

**Comment by qiaohaijun** [comment]
I got dropout ops convert fail in tensorrt 4.0
```
2018-08-30 11:11:20.711520: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:418] subgraph conversion error for subgraph_index:10 due to: "Unimplemented: Require 4 dimensional input. Got 3 query_lookup/Dropout/cond/dropout/Shape/Switch" SKIPPING......( 3 nodes)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @miaout17: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @miaout17: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @miaout17: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @miaout17: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @miaout17: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by petewarden** [comment]
I've just approved the PR related to this one. Sorry for the delay in reviewing.

**Comment by nanshihui** [comment]
which version can solve this problem。hope comming soon

**Comment by mohapatras** [comment]
Any update on the solution yet ?

**Comment by kmonachopoulos** [comment]
I think that the only way to remove dropout for now is to comment out the dropout lines in your python source code and export the model without them ...

**Comment by danvargg** [comment]
Any update on this issue yet?

**Comment by zh794390558** [comment]
Any update on this issue yet?

**Comment by danvargg** [comment]
> There is also this SO (attempt at a) solution: http://stackoverflow.com/questions/40358892/wipe-out-dropout-operations-from-tensorflow-graph
> and my attempt at this:
> 
> ```python
> for node in temp_graph_def.node:
>     for idx, i in enumerate(node.input):
>         input_clean = node_name_from_input(i)
>         if input_clean.endswith('/cond/Merge') and input_clean.split('/')[-3].startswith('dropout'):
>             identity = node_from_map(input_node_map, i).input[0]
>             assert identity.split('/')[-1] == 'Identity'
>             parent = node_from_map(input_node_map, node_from_map(input_node_map, identity).input[0])
>             pred_id = parent.input[1]
>             assert pred_id.split('/')[-1] == 'pred_id'            
>             good = parent.input[0]
>             node.input[idx] = good
> ```

Hi @cancan101, what is `node_name_from_input`? 
I'm trying to implement this manual fix for a conversion `Tensorflow > CoreML`.

**Comment by beniroquai** [comment]
One comment to add - it's hard to debug if you don't know what this means: 
```Here is a list of operators for which you will need custom implementations: RandomUniform.````
or 
```Exception: Placeholder Placeholder should be specied by input_arrays.````
which either indicate the use for a ```tf.placeholder``` for the dropout probability or the dropout node in the ```tf.lite```conversion process. Outcommenting as suggested helped!
